swarm/chunk/tags.go

```diff
@@ -0,0 +1,211 @@
+package chunk
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+var (
+	errExists = errors.New("already exists")
+	errNoETA  = errors.New("unable to calculate ETA")
+)
+
+// tags holds the tag infos indexed by name
+type tags struct {
+	tags *sync.Map
+	rng  *rand.Rand
+}
+
+// NewTags creates a tags object
+func newTags() *tags {
+
+	return &tags{
+		tags: &sync.Map{},
+		rng:  rand.New(rand.NewSource(time.Now().Unix())),
+	}
+}
+
+// New creates a new tag, stores it by the name and returns it
+// it returns an error if the tag with this name already exists
+func (ts *tags) New(s string, total int) (*Tag, error) {
+	t := &Tag{
+		uid:       ts.rng.Uint32(),
+		name:      s,
+		startedAt: time.Now(),
+		total:     uint32(total),
+	}
+	_, loaded := ts.tags.LoadOrStore(s, t)
+	if loaded {
+		return nil, errExists
+	}
+	return t, nil
+}
+
+// Inc increments the state count for a tag if tag is found
+func (ts *tags) Inc(s string, f State) {
+	t, ok := ts.tags.Load(s)
+	if !ok {
+		return
+	}
+	t.(*Tag).Inc(f)
+}
+
+// Get returns the state count for a tag
+func (ts *tags) Get(s string, f State) int {
+	t, _ := ts.tags.Load(s)
+	return t.(*Tag).Get(f)
+}
+
+// State is the enum type for chunk states
+type State = uint32
+
+const (
+	SPLIT  State = iota // chunk has been processed by filehasher/swarm safe call
+	STORED              // chunk stored locally
+	SEEN                // chunk previously seen
+	SENT                // chunk sent to neighbourhood
+	SYNCED              // proof is received; chunk removed from sync db; chunk is available everywhere
+)
+
+// Tag represents info on the status of new chunks
+type Tag struct {
+	uid       uint32    // a unique identifier for this tag
```

Comment:
needed?

Reply:
how are you going to identify the tag with a unique id in the `sync.Map`?